appserver/payara-appserver-modules/microprofile/fault-tolerance/src/main/java/fish/payara/microprofile/faulttolerance/policy/FaultTolerancePolicy.java

```diff
@@ -0,0 +1,609 @@
+ * "Portions Copyright [year] [name of copyright owner]"
+ *
+ * Contributor(s):
+ * If you wish your version of this file to be governed by only the CDDL or
+ * only the GPL Version 2, indicate your decision by adding "[Contributor]
+ * elects to include this software in this distribution under the [CDDL or GPL
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.microprofile.faulttolerance.policy;
+
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.time.Duration;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.interceptor.InvocationContext;
+
+import org.eclipse.microprofile.faulttolerance.exceptions.BulkheadException;
+import org.eclipse.microprofile.faulttolerance.exceptions.CircuitBreakerOpenException;
+import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceDefinitionException;
+import org.eclipse.microprofile.faulttolerance.exceptions.FaultToleranceException;
+import org.eclipse.microprofile.faulttolerance.exceptions.TimeoutException;
+
+import fish.payara.microprofile.faulttolerance.FaultToleranceConfig;
+import fish.payara.microprofile.faulttolerance.FaultToleranceService;
+import fish.payara.microprofile.faulttolerance.FaultToleranceMetrics;
+import fish.payara.microprofile.faulttolerance.state.BulkheadSemaphore;
+import fish.payara.microprofile.faulttolerance.state.CircuitBreakerState;
+
+/**
+ * The {@link FaultTolerancePolicy} describes the effective aggregated policies to use for a particular {@link Method}
+ * when adding fault tolerant behaviour to it.
+ * 
+ * The policies are extracted from FT annotations and the {@link FaultToleranceConfig}.
+ * 
+ * In contrast to the plain annotations the policies do consider configuration overrides and include validation of the
+ * effective values.
+ * 
+ * The policy class also reduces the need to analyse FT annotations for each invocation and works as a consistent source
+ * of truth throughout the processing of FT behaviour that is convenient to pass around as a single immutable value.
+ *
+ * @author Jan Bernitt
+ */
+public final class FaultTolerancePolicy implements Serializable {
```

Comment:
Web filters are (or should be) semantically independent and therefore should be elements in a chain that do not know about each other and naturally the code is kept separate as those are independent functionalities. 

The stages of the fault tolerance behaviour are not like that. They are semantically linked and dependent and work as a union even though they represent concepts that can be used individually. As soon as multiple annotations are used there are semantics on how this combination behaves wherefore these are stages of a single mechanism. Pulling it apart only tricks the reader into thinking that those are independent when they are not. Most of the time you'll debug combinations and make changes that affect multiple stages to make the overall 
behaviour correct. Therefore I think its better to have these in the same file - even if it gets longer - because semantically this is one complex mechanism with multiple aspects.

I did them default visible is to allow to unit test intermediate stages directly.